openfl-tutorials/interactive_api/PyTorch_MVTec_PatchSVDD/workspace/PatchSVDD_with_Director.ipynb

```diff
@@ -0,0 +1,1170 @@
+    "        h = self.fc3(h)\n",
+    "        return h\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "db853fe0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MyEnsembledModel(nn.Module):\n",
+    "    def __init__(self, modelA, modelB, modelC):\n",
+    "        super().__init__()\n",
+    "        self.modelA = modelA\n",
+    "        self.modelB = modelB\n",
+    "        self.modelC = modelC\n",
+    "        \n",
+    "    def forward(self, x1, x2,x3):\n",
+    "        x1 = self.modelA(x1)\n",
+    "        x2 = self.modelB(x2)\n",
+    "        x3 = self.modelC(x3)\n",
+    "        x = torch.cat((x1, x2,x3), dim=1)\n",
+    "        return x\n",
+    "\n",
+    "D = args['D']\n",
+    "lr = float(args['lr'])\n",
+    "enc = EncoderHier(64, D).cuda()\n",
+    "cls_64 = PositionClassifier(64, D).cuda()\n",
+    "cls_32 = PositionClassifier(32, D).cuda()\n",
+    "\n",
+    "\n",
+    "model = MyEnsembledModel(enc, cls_64, cls_32)\n",
+    "modules = [enc, cls_64, cls_32]\n",
+    "params = [list(module.parameters()) for module in modules]\n",
+    "params = reduce(lambda x, y: x + y, params)\n",
+    "optimizer_adam = torch.optim.Adam(params=params, lr=lr)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "caroline-passion",
+   "metadata": {},
+   "source": [
+    "#### Register model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "handled-teens",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from copy import deepcopy\n",
+    "\n",
+    "framework_adapter = 'openfl.plugins.frameworks_adapters.pytorch_adapter.FrameworkAdapterPlugin'\n",
+    "MI = ModelInterface(model=model, optimizer=optimizer_adam, framework_plugin=framework_adapter)\n",
+    "\n",
+    "# Save the initial model state\n",
+    "initial_model = deepcopy(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "portuguese-groove",
+   "metadata": {},
+   "source": [
+    "### Define and register FL tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "increasing-builder",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TI = TaskInterface()\n",
+    "import torch\n",
+    "import tqdm\n",
+    "from utils import cnn_output_size\n",
+    "\n",
+    "# The Interactive API supports registering functions definied in main module or imported.\n",
+    "def function_defined_in_notebook(some_parameter):\n",
+    "    print(f'Also I accept a parameter and it is {some_parameter}')\n",
+    "\n",
+    "# Task interface currently supports only standalone functions.\n",
+    "@TI.add_kwargs(**{'some_parameter': 42})\n",
+    "@TI.register_fl_task(model='model', data_loader='train_loader', \\\n",
+    "                     device='device', optimizer='optimizer')     \n",
+    "\n",
+    "def train(model, train_loader, optimizer, device, some_parameter=None):\n",
+    "    if not torch.cuda.is_available():\n",
```

Comment:
This example doesn't work if the director and the workspace are located on different services or in different virtual environments, because serialisable objects use requirements  (torch, ngtpy, at least) that are missing on the director.

Reply:
@aleksandr-mokrov how should I ensure this?